Интеграция с вики

Pos13 opened the pull request:
**Внимание! Для корректной работы в файле конфигурации _config\config.txt_ должен быть прописан параметр WIKIURL:**
`## Wiki URL`
`WIKIURL https://wiki.ss13.ru`

### Изменения в игре
1. **Библиотекарь получает возможность распечатывать некоторые статьи из вики.**
Данная возможность работает по аналогии с печатью обычных книг: сначала консоль печатает "бумагу", которая играет роль заготовки для будущей книги; потом переплётная машина позволяет превратить её в "книгу". Статьи, доступные для печати, регулируются вне кода, вот этим списком: [Available in library](https://wiki.ss13.ru/index.php?title=Категория:Available_in_library).
_Заготовка:_
![image](https://user-images.githubusercontent.com/64479446/85908299-3a4e6500-b81d-11ea-9111-ff2c286ce03f.png)
_Готовая книга:_
![image](https://user-images.githubusercontent.com/64479446/85908311-44706380-b81d-11ea-9ba7-ab593d0da584.png)
Вы могли не знать, но, если расположить переплётную машину рядом с консолью, то обычные книги будут печататься сразу в ней, мгновенно. Это правило выполняется и при распечатке статей из вики.

2. **Емагнутая консоль открывает доступ к дополнительным статьям.**
Если быть точным, то к статьям из этого списка: [Available in hacked library](https://wiki.ss13.ru/index.php?title=Категория:Available_in_hacked_library).
Более того, содержимое некоторых статей будет расширено. (Точнее, содержимое некоторых статей, распечатанных на _невзломанном_ компьютере, будет частично скрыто. Подробности ниже, в разделе "Изменения в вики")

3. **Владельцы ноутбуков и планшетов тоже получают возможность печатать статьи, но с оговорками.**
Можно скачать программу "Knowledge base", аналогичную программе на компьютере библиотекаря. Но, при печати с помощью встроенного принтера (если он есть), получается неудобная для чтения заготовка. Её можно превратить в полноценную книгу только в переплётной машине.
Как вариант, владелец портативного компьютера может просто встать рядом с переплётной машиной - книги сразу будут печататься в ней.
Емагнутые портативные компьютеры работают с программой аналогично емагнутой консоли.
![image](https://user-images.githubusercontent.com/64479446/85908930-04f74680-b820-11ea-805e-7ea877114aaa.png)
4. **Отдельный класс для вики-книг.**
Хотите создать инструкцию для ~~бес~~толковых работников? Нет ничего проще:

`/obj/item/weapon/book/wiki/alarms`
       **`topic = "Standard Operating Procedure"`**
       `title = "What to do then everything is bad and you are the boss"`
       `icon_state = "bookSolGovLaw"`

Такая книга при открытии автоматически загрузит содержимое статьи "Standard Operating Procedure".

"Книжное представление" вики-статьи не содержит никаких ссылок, вообще. Все элементы вики-интерфейса срезаны. Так что "сидеть в интернете через спейслоу" больше не выйдет. Дополнительный бонус: все загрузки и парсинг осуществляются на клиенте, с помощью _javascript_ и _MediaWiki API_.
![image](https://user-images.githubusercontent.com/64479446/85909676-b21f8e00-b823-11ea-8257-8cf7288e53b4.png)

5. **Заменены встроенные в код мануалы.**
Вместо устаревших англоязычных мануалов с захардкоженным текстом теперь книги, подгружающие соответствующий контент из вики. Иконки, заголовки и описания не пострадали, внешне всё по-прежнему. В основном замены достаточно прямолинейные (вроде _Barman Recipes_, открывающего [Guide to Drinks](https://wiki.ss13.ru/index.php?title=Guide_to_Drinks)). Вот некоторые особенные случаи:

- Для мануалов, не имеющих аналогов, созданы соответствующие статьи на вики (см. [Diona Harvest](https://wiki.ss13.ru/index.php?title=Diona_Harvest) и [Brief Chemist Guide](https://wiki.ss13.ru/index.php?title=Brief_Chemist_Guide));
- Мануалы о клонировании и криосне вырезаны;
- Россыпь бесполезных микро-книг на высшем ~~эльфийском~~ английском о перспективах и прикладных применениях науки исследования аномалий заменена на [Guide to Xenoarchaeology](https://wiki.ss13.ru/index.php?title=Guide_to_Xenoarchaeology).

### Изменения в вики
1. **Категория статей, доступных для печати в игре:  [Available in library](https://wiki.ss13.ru/index.php?title=Категория:Available_in_library).**
Добавление или удаление статьи из этой категории автоматически разрешит или запретит её печать в игре.
Для регулирования статей, доступных для печати с емагнутой консоли, есть категория [Available in hacked library](https://wiki.ss13.ru/index.php?title=Категория:Available_in_hacked_library).
Чтобы добавить статью в категорию, откройте её на вики, нажмите "править" и добавьте в конец текста [[Category:Available in library]].
2. **Шаблон {{Ingame}}.**
![image](https://user-images.githubusercontent.com/64479446/85910505-eac16680-b827-11ea-88f0-84094357f016.png)
- **Статья без этой плашки будет недоступна из игры**;
- Название книги, заданное в плашке, переносится в игру (применяется при первом открытии книги);
- Весь текст, расположенный выше плашки, вырезается. Это позволяет скрыть нежелательные части (другие плашки или вступительные слова вроде "Не стоит пытаться играть завхозом, не приобретя для начала опыт работы грузчиком, иначе у вас есть шанс запороть раунд.");
- Плашка побуждает редакторов осторожнее выбирать выражения.

~~_Примечание_: в коде предусмотрена, но выключена возможность запретить просмотр страниц, не помеченных такой плашкой.~~

Более подробно использование этой плашки описано в её статье: [Шаблон:Ingame](https://wiki.ss13.ru/index.php?title=Шаблон:Ingame).

3. **Шаблон {{NT Censorship|}}**
![image](https://user-images.githubusercontent.com/64479446/85911190-5279b080-b82c-11ea-8e5e-23384d773d6f.png)
В такой шаблон включается всё, что должно избежать взгляда верного корпоративного работника. Содержимое становится видимым только при печати со взломанного компьютера. Рамочку, впрочем, видят все - пусть игроки видят, что здесь прячется что-то интересное!
![image](https://user-images.githubusercontent.com/64479446/85911300-2579cd80-b82d-11ea-9ec8-3ca5c59f165b.png)

Более подробно использование этой плашки описано в её статье: [Шаблон:NT_Censorship](https://wiki.ss13.ru/index.php?title=Шаблон:NT_Censorship).

Известные баги:
- Теоретически, при нестабильном соединении книга может не открыться (наблюдалось 1 раз из бесчисленного множества тестов). В таком случае будет выведено предусмотренное кодом сообщение об ошибке.
- Содержание в статьях отключено, ибо якорные ссылки не работают в BYOND. Вообще никакие, ни при каких обстоятельствах. Причём раньше, возможно, работали, но сломались при апдейте, который игрался с [Winset and Winget via JavaScript](http://www.byond.com/docs/ref/#/{skin}/control/browser).
- Иногда небольшие картинки в преамбуле заготовок под книги (_"This is a book template..."_) не отображаются. Это видно на первой картинке в этом ПР. Иногда отображаются, иногда - нет. С чем связано - не знаю.

close #2674 #2799 #2780

- [x] Pull Request полностью завершен, мне не нужна помощь чтобы его закончить.
- [x] Я внимательно прочитал все свои изменения и багов (кроме упомянутых) в них не нашел.
- [x] Я запускал сервер со своими изменениями локально и все протестировал.
- [x] Я ознакомился c [Guide to Contribute](https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md).

Pos13 committed:
rework(manuals): Hardcode -> Wiki

Готовые книги-руководства изменены так, чтобы вместо захардкоженного текста выводить странички из игровой вики.

Pos13 committed:
add(knowledge base): Wiki library program

Добавлена возможность распечатывать тексты из вики в библиотеке.
Исправлен мелкий баг, съедающий заголовок текста при распечатке.

Pos13 committed:
update(library): wiki titles

Перенос заголовка книги из плашки {{Ingame}} в игру.

Pos13 committed:
Revert "update(library): wiki titles"

This reverts commit 9b2c3550ead8f0a577b0bfc30b54b70c8d8f8023.

Pos13 committed:
update(wiki book): title -> name

Pos13 committed:
update(example config): WIKIURL

Записал изменения в example\config.txt.

Doster-d commented:
Хочу, хочу, хочу! Это мастхев!

ArmolitskiyCarpoeb commented:
Просто. Ахуенно. Отличная работа!

Doster-d commented:
Дорогие старые друзья, помните этот ПР

Epicus7 commented on code/modules/modular_computers/file_system/programs/generic/library.dm:
```diff
@@ -227,3 +223,58 @@ The answer was five and a half years -ZeroBits
+				return 1
+
+		// Regular print (creates book template)
+		if(istype(nano_host(), /obj/item/modular_computer) && !nano_host():nano_printer)
```
И тут двоеточие.

Pos13 replied:
Исправил

Pos13 committed:
Update(bookshelf)

changed 'for' cycle

Co-authored-by: Epicus <epicus7@gmail.com>

Pos13 committed:
Update(bookshelf)

Скобки

Co-authored-by: Epicus <epicus7@gmail.com>

Pos13 commented on code/modules/library/lib_items.dm:
```diff
@@ -224,12 +227,78 @@
+	var/style = WIKI_MINI
+	var/censored = 1
+
+/obj/item/weapon/book/wiki/Initialize(mapload, ntopic, ncensored, nstyle)
```
> Что это за странная n в начале аргументов?

Чтобы избежать путаницы. Я уже видел такое где-то в коде.

> Алсо что это вообще такое, откуда вызывается Initialize и откуда берутся аргументы?

Я просто следовал [рекомендации из Guide to Contribute](https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md#%D0%BF%D1%80%D0%B5%D0%B4%D0%BF%D0%BE%D1%87%D0%B8%D1%82%D0%B0%D0%B9%D1%82%D0%B5-initialize-%D0%B2%D0%BC%D0%B5%D1%81%D1%82%D0%BE-new-%D0%B4%D0%BB%D1%8F-atom-%D0%BE%D0%B1%D1%8A%D0%B5%D0%BA%D1%82%D0%BE%D0%B2-%D1%80%D0%B0%D0%B7%D0%BC%D0%B5%D1%89%D0%B0%D0%B5%D0%BC%D1%8B%D1%85-%D0%BD%D0%B0-%D0%BA%D0%B0%D1%80%D1%82%D0%B5). Аргументы вполне корректно передаются при вызове _new_. Объекты, изначально размещённые на карте, тоже выглядели созданными верно.

Epicus7 replied:
Покажи, пожалуйста, место, где ты создаешь книги с этими четыремя аргументами. Я не нашел.

Pos13 replied:
**lib_machines.dm**:
`/obj/machinery/bookbinder/proc/print_wiki(topic, censorship)`
`      var/obj/item/weapon/book/wiki/book`
`      if(topic in GLOB.premade_manuals)`
`            var/manual_type = GLOB.premade_manuals[topic]`
**`            book = new manual_type(src.loc, topic, censorship)`**
`      else`
**`            book = new /obj/item/weapon/book/wiki(src.loc, topic, censorship, WIKI_MINI)`**
`            book.icon_state = "book[rand(1,7)]"`
`      if(book.layer <= src.layer)`
`            book.layer = src.layer + 0.01 // Prevents spawning under bookbinder`
`      return book`
**library.dm**:
`/datum/nano_module/wiki/Topic(href, href_list)`
...
`// Regular print (creates book template)`
`if(istype(nano_host(), /obj/item/modular_computer) && !nano_host():nano_printer)`
`      to_chat(usr, SPAN_DANGER("Error: No printer detected. Unable to print document."))`
`      return 1`
**`new /obj/item/weapon/book/wiki/template(get_turf(nano_host()), href_list["topic"], emagged ? 0 : 1)`**
`return 1`

Pos13 replied:
Кстати, попробовал использовать **src.censored = censored** вместо использования **ncensored**, но получил странное поведение. Что было не так - не знаю, упорствовать не стал.

Epicus7 replied:
Прикольно, Initialize вызывается из New со всеми аргументами. Я даже не знал про это.

Ладно, пофиг.

Pos13 committed:
Update(book/wiki)

Added TODO

Co-authored-by: Epicus <epicus7@gmail.com>

Pos13 commented on code/modules/library/manuals.dm:
```diff
@@ -0,0 +1,178 @@
+	topic = "Brief Chemist Guide"
+	style = WIKI_MOBILE
+
+// -----
+// Books below are not premade manuals. They just hold custom settings to display wiki page properly
+// -----
```
Сверху - заранее описанные в коде книги, которые предназначены для размещения на карте. С прописанным заголовком, автором, собственной иконкой.
Снизу - книги, у которых нет отличительных черт, но для которых после печати желательно поменять параметры - ширину окна или стиль текста. Для читаемости.
Для книг, не имеющих особенностей, подклассы не требуются.

Pos13 commented on code/modules/library/lib_machines.dm:
```diff
@@ -67,17 +67,41 @@
+	else
+		book = new /obj/item/weapon/book/wiki(src.loc, topic, censorship, WIKI_MINI)
+		book.icon_state = "book[rand(1,7)]"
+	if(book.layer <= src.layer)
+		book.layer = src.layer + 0.01 // Prevents spawning under bookbinder
```
Все машины в библиотеке, включая переплётную машину, консоль и даже мусорку, лежат на слое OBJ_LAYER (layer=3). Вообще, если ткнуть на станции в случайную машину, она, скорее всего, лежит на слое 3. Короче, я не менял слой переплётной машины по-умолчанию, так как не знаю, почему для всех **/obj/machinery** он равен 3.

Pos13 replied:
Можно сразу прописать **layer = BELOW_OBJ_LAYER** либо для всех машин, либо только для переплётной. Но я не знаю, не сломается ли что-нибудь.

Pos13 replied:
Короче, задал всем машинам по дефолту **layer = BELOW_OBJ_LAYER**, не стал мелочиться. На вид всё нормально, мусорки из экрана в лицо не выпрыгивают.

Pos13 commented on code/modules/modular_computers/file_system/programs/generic/library.dm:
```diff
@@ -145,14 +145,10 @@ The answer was five and a half years -ZeroBits
-				bndr.visible_message("\The [bndr] whirs as it prints and binds a new book.")
+				var/obj/item/weapon/book/new_book = bndr.print(current_book["content"], current_book["title"], current_book["author"])
+				if(new_book)
+					new_book.desc = current_book["author"]+", "+current_book["title"]+", "+"USBN "+current_book["id"]
```
Копипаста.

Pos13 committed:
Update(print book)

Spaces.

Co-authored-by: Epicus <epicus7@gmail.com>

Pos13 commented on code/modules/modular_computers/file_system/programs/generic/library.dm:
```diff
@@ -227,3 +223,58 @@ The answer was five and a half years -ZeroBits
+/datum/nano_module/wiki/ui_interact(mob/user, ui_key = "main", datum/nanoui/ui = null, force_open = 0, datum/topic_state/state = GLOB.default_state)
+	var/list/data = host.initial_data()
+
+	var/emagged = host && (istype(host, /obj/item/modular_computer) || istype(host, /datum/computer_file/program)) ? host:computer_emagged : 0
```
> Какой-то синтаксический ад. Во-первых, host ты выше используешь, так что он не может быть null, иначе у тебя рантайм раньше случится.

Уберу проверку.

> Во-вторых, опять ты используешь запрещенный оператор : .

Его использование оправдано. У **/obj/item/modular_computer** и **/datum/computer_file/program** есть переменная с одинаковым названием **computer_emagged**, играющая одну и ту же функцию.

> В-третьих, вместо 0 должен быть FALSE здесь.

Переменная затем используется в коде на javascript. Я хочу быть точно уверен, что там 0 или не 0.

Pos13 replied:
Исправил.
Заменил одну красивую строчку на семь обычных.
Грустное двоеточие не любит тавтологию и грустит -> **:(**

Pos13 committed:
update(books)

Various fixes. В основном синтаксическое.
Все /obj/machinery перемещены на layer = BELOW_OBJ_LAYER.

Pos13 commented on code/modules/library/lib_items.dm:
```diff
@@ -147,6 +149,7 @@
 	var/title		 // The real name of the book.
 	var/carved = 0	 // Has the book been hollowed out for use as a secret storage item?
 	var/obj/item/store	//What's in the book?
+	var/width = 650
```
Сделал

Pos13 commented on code/modules/library/lib_items.dm:
```diff
@@ -159,7 +162,7 @@
 			return
 	if(src.dat)
-		user << browse(dat, "window=book;size=1000x550")
+		user << browse(dat, "window=book_[title];size=[width]x650")
```
Теперь по дефолту **title = "Untitled"**

Pos13 commented on code/modules/library/lib_items.dm:
```diff
@@ -224,12 +227,78 @@
 		user.visible_message("<span class='notice'>You open up the book and show it to [M]. </span>", \
 			"<span class='notice'> [user] opens up a book and shows it to [M]. </span>")
-		M << browse("<i>Author: [author].</i><br><br>" + "[dat]", "window=book;size=1000x550")
+		M << browse(dat, "window=book_[title];size=[width]x650")
```
Теперь по дефолту **title = "Untitled"**

Pos13 commented on code/modules/library/lib_machines.dm:
```diff
@@ -67,17 +67,41 @@
+		if(istype(O, /obj/item/weapon/paper))
+			print(O:info, "Print Job #" + "[rand(100, 999)]")
+		if(istype(O, /obj/item/weapon/book/wiki/template))
+			print_wiki(O:topic, O:censored)
```
Убрал. **O:info**, кстати, было в старом коде.

Pos13 commented on code/modules/modular_computers/file_system/programs/generic/library.dm:
```diff
@@ -227,3 +223,58 @@ The answer was five and a half years -ZeroBits
 	if(query.Execute())
 		log_and_message_admins("has deleted the book: \[[sqlid]\] \"[title]\" by [author]", user)
+
+#define WIKI_COMMON_CATEGORY "Available_in_library"
+#define WIKI_HACKED_CATEGORY "Available_in_hacked_library"
```
Сделал.

Pos13 commented on code/modules/modular_computers/file_system/programs/generic/library.dm:
```diff
@@ -227,3 +223,58 @@ The answer was five and a half years -ZeroBits
+	if(..())
+		return 1
+	if(href_list["topic"])
+		var/emagged = host && (istype(host, /obj/item/modular_computer) || istype(host, /datum/computer_file/program)) ? host:computer_emagged : 0
```
Исправил

MemoryDreams commented:
Это большая ошибка. Не надо. Не делайте этого. Давайте сделаем нормальные гайды, нв которые игроки смогут ссылаться в диалогах, а не вики странички, которые никто и никогда не будет открывать в игре.
Одумайтесь. Ими никто, никогда не пользовался. Это большая ошибка. Особенно с учётом того, что сента принял решение об ограничении информации на вики. Это просто не имеет смысла.
Это большая ошибка и я хочу наложить на это вето.

Pos13 commented:
> Это большая ошибка. Не надо. Не делайте этого. Давайте сделаем нормальные гайды, нв которые игроки смогут ссылаться в диалогах, а не вики странички, которые никто и никогда не будет открывать в игре.
> Одумайтесь. Ими никто, никогда не пользовался. Это большая ошибка. Особенно с учётом того, что сента принял решение об ограничении информации на вики. Это просто не имеет смысла.
> Это большая ошибка и я хочу наложить на это вето.

Нормальные гайды теперь можно будеть писать с помощью вики. Не нужно будет быть кодером, чтобы поменять или внести текст своей книги (как было раньше с захардкоженными руководствами)
К тому же, содержимое таких вики-книг не похоже на интернет-страницу, и её название можно регулировать через плашку {{Ingame}}. Короче, нет проблемы упомянуть такую в разговоре.

MemoryDreams commented:
> > Это большая ошибка. Не надо. Не делайте этого. Давайте сделаем нормальные гайды, нв которые игроки смогут ссылаться в диалогах, а не вики странички, которые никто и никогда не будет открывать в игре.
> > Одумайтесь. Ими никто, никогда не пользовался. Это большая ошибка. Особенно с учётом того, что сента принял решение об ограничении информации на вики. Это просто не имеет смысла.
> > Это большая ошибка и я хочу наложить на это вето.
> 
> Нормальные гайды теперь можно будеть писать с помощью вики. Не нужно будет быть кодером, чтобы поменять или внести текст своей книги (как было раньше с захардкоженными руководствами)
> К тому же, содержимое таких вики-книг не похоже на интернет-страницу, и её название можно регулировать через плашку {{Ingame}}. Короче, нет проблемы упомянуть такую в разговоре.

Внутреигровые книги и гайды на вики написаны на разных языках. Их нельзя мешать. Это будет ломать атмосферу.

MrMagick2104 commented:
> Внутреигровые книги и гайды на вики написаны на разных языках. Их нельзя мешать. Это будет ломать атмосферу.

Те статьи, что попадают в игру, должны быть изначально переделаны и на них должна быть плашка {{Ingame}}.
При этом, ООЦ-материалы из статьи, если помечены, в книге показываться _не будут_, но в интернет-статье видны _будут_. В чём проблема-то?

MemoryDreams commented:
Это большая ошибка.

MrMagick2104 commented:
Аргументируй.

MemoryDreams commented:
It feels wrong. Идеальным решением было бы склонить игроков к написанию собственных гайдов, сильно ограниченных по полноте информации, но дающие разные перспективы на разные проблемы. Унифицированные гайды убирают всякую нужду в разбирательстве что и как работает. Я считаю, что на вики должна быть только информация об управлении и всяких условностях, которые просто нельзя рассказать непосредственно в игре. 
Унифицированные гайды это от лукавого и я против такого решения.

TobyThorne commented:
> склонить игроков к написанию собственных гайдов

Аха. У нас половина игроков и три слова без ошибок написать не могут, а ты про запиливание гайдов.

MemoryDreams commented:
> Аха. У нас половина игроков и три слова без ошибок написать не может, а ты про запиливание гайдов.

Пока есть игроки, которые готовы заниматься вики, гайды писать будет кому.

MrMagick2104 commented:
> склонить игроков к написанию собственных гайдов

Игроки это никогда не будут делать. Почему? Да просто потому, что они в любом случае создают небольшие или большие объединения по знаниям и обмениваются ими, неизбежно. Или берут идеи от других. Или ещё что-то.
И делают они без гайдов. Просто потому что обмен информации в игре или ООЦ - намного быстрее и проще, к нужным людям, а писать гайды никому не приелось, кроме особых людей - викиврайтеров. Не могу их судить, зачем они это делают, но, однозначно, это - полезно, если информация в них верная **И** гайды написаны максимально общим языком.
Такой гайд понадобится в виде книжки уборщику, который решил настроить суперматирию сам или что-то такое, просто чтобы станция не сидела без света, а он - единственный, кто может это сделать, хотя он и не умеет это делать ни в ООЦ, ни в ИС, возможно.
И гайд с вики ему _должен_ помочь в этом, потому что это - задача гайда с вики - помогать новичкам, как ИЦ, так и ООЦ.
А если гайды будут писать игроки, то выйдет херня в любом случае:
-Они не будут писать гайды, это раз, т.к. если бы писали, то и на вики бы у нас тоже активности было больше, чем от трёх людей максимум
-Какой гайд положить рядом с суперматерией, чтобы уборщик мог прийти, прочитать, кое-ка настроить? Чьего авторства? Он должен быть максимально простой, понятный и эффективный.
Если всем вместе решать, то тогда это будет избыточность, ведь можно просто подлатать статью на вики так, чтобы она удовлетворяла условиям для гайда выше, после чего портировать её на станцию.
-Даже если вот эту конкретную фичу запилят, что мешает игрокам пилить гайды в IC, либо пилить их в OOC на вики, создавая новые книжки, доступные игрокам? Я думаю, если в сента пропихнуть идею, то можно вполне себе добавить возможность создания таких "коммьюнити-мейд" гайдов, с некоторой цензурой со стороны сената/хедвики (которого у нас нет?)
-Продолжая случай с суперматерией: большинство инженеров, если и будут, то будут делать гайды _непонятными_ простому уборщику или новичку, из-за чего он материю может настроить _неправильно_, вместо среднего гайда.
-Если говорить о том, что унифицированные гайды - плохо, потому что игроки будут тупо следовать по нему, то это будет ложью. Большинство тех же инженеров от самого начала и до самого конца забивало на гайд суперматерии, где ни слова не было сказано ни о коллекторах в реакторе, ни о фороне, etc.
Все так всё равно делали, даже несмотря на в своё время 'outdated' гайд по суперматерии.
Зачастую, такие гайды для новичков, а не для большинства.


В чём проблема давать возможность давать игрокам общие гайды, хорошо реализованные как тут, в IC и OOC, а также дать им возможность писать свои гайды (которая уже есть, ага)?
Только тут будет преимущество в том, что уборщик будет тупить не в браузер в альт-табе, а в книжку в игре, что лучше, как минимум, по двум причинам:
-Это логичнее с точки зрения игрового мира, знания у уборщика не появляются от Аллаха, а из книжки.
-Альттабаться неудобно в любом случае, если у человека не два монитора, то он полностью отстраняется от мира вокруг.

А ещё это уже закожено и имплиментировано, потому, почему нет? От этого только плюсы ведь.

> Пока есть игроки, которые готовы заниматься вики, гайды писать будет кому.

А ты можешь сказать, сколько людей занимается вики сейчас, в сравнении с ежедневным онлайном? А ещё, кто у нас хедвики?

MemoryDreams commented:
> А если гайды будут писать игроки, то выйдет херня в любом случае

О чём ты говоришь? Их всегда писали и пишут до сих пор игроки. 
Проблема стены текста, которую ты написал в том, что ты всё время ссылаешься на пример с суперматерией, тогда как она является скорее исключением (даже в правилах уборщик настраивающий суперматерию вынесен как исключение). Нет ни одного случая кроме суперматерии, когда такие гайды будут к месту, потому что ни один гайд кроме суперматерии не является необходимым для обеспечения нормальной работы станции. Возьмём, к примеру, химию или ксенобиологию. Максимум информации, который можно поместить в такой гайд для этих отделов, это описания, какие аппараты для чего нужны. Но вот вопрос, если игрок вступает в раунд химиком или ксенобиологом, не зная, что гриндер нужен для измельчения твёрдых вещей, например, то возникает вопрос, а что он забыл на этой профессии? В то же самое время, любой другой член экипажа, ассистент, например, просто ничего не сможет сделать без знаний конкретных рецептов. Ну и зачем это нужно, спрашивается? 

Да и даже в случае с материей, этими гайдами просто не будут пользоваться. Вспомни себя в первые недели игры. Ты эти книжки либо не замечал, либо открывал единожды и со словами "Ага, это кусочек вики. Пойду лучше полную версию открою" закрывал её, больше никогда не вспоминая о такой фичи. 

> Даже если вот эту конкретную фичу запилят, что мешает игрокам пилить гайды в IC, либо пилить их в OOC на вики, создавая новые книжки, доступные игрокам?

А зачем им пилить гайды, когда уже есть "золотой стандарт" в виде официальных гайдов? Такие вещи очень демотивируют заниматься и без того не самым благодарным занятием. Что нужно сделать, так это завлечь игроков к тому, чтобы они делились опытом. Провести какой-нибудь конкурс или вроде того. Но никак не заменять всё бездушным копированием вики.

MrMagick2104 commented:
> Нет ни одного случая кроме суперматерии, когда такие гайды будут к месту, потому что ни один гайд кроме суперматерии не является необходимым для обеспечения нормальной работы станции.

Да, но:
-Во-первых, это уже в _тысячи_ раз лучше, чем то, что есть сейчас. Это - удобная система для создания гайдов внутри игры без необходимости хардкодить их в игру, тем самым позволяя легко модифицировать их.
Куда проще сделать саму статью на вики при помощи удобного викидота или что у вас там, добавить картиночки, подогнать под шаблон, после посмотреть на результат, а только уже потом добавлять в игру, при этом без необходимости поднимать локалочку и качать гитхаб десктоп или бранч каждый раз, чтобы посмотреть на все изменения в процессе создания статьи, а потом просто добавить шесть строчек и гайд в билде.
Не обязательно даже портировать гайды в игру. Сама по себе система от ОП-а намного проще и удобнее для большинства пользователей и людей, которые будут делать гайды в будущем.
-Во-вторых, если есть прямая интеграция с вики, то это уже удобно для пользователя. Ему не нужно идти в браузер, открывать вики и искать там материал, когда он вполне себе может взять книжку из отдела и посмотреть нужный ему предмет.
Бывают просто такие ситуации, что забыл последовательность действий, допустим, сбора меха на робототехнике, либо её ещё не выучил, либо какого-нибудь другого предмета, где нужна чёткая последовательность действий.
Куда удобнее открыть книжку, валяющуюся рядом, в случае с робототехником, нежели искать гайд ту роботикс в интернете.

> Вспомни себя в первые недели игры. Ты эти книжки либо не замечал, либо открывал единожды и со словами "Ага, это кусочек вики. Пойду лучше полную версию открою" закрывал её, больше никогда не вспоминая о такой фичи.

В первый раз я материю сам настраивал с каким-то таким же гением, по книжке, а не по вики. 
И гайды по робототехнике той же, на первых неделях, я читал не только на вики, но и в книжках. Удивительно, но там тогда информации было больше, чем на вики.
Аргумент - инвалид.

> Возьмём, к примеру, химию или ксенобиологию. Максимум информации, который можно поместить в такой гайд для этих отделов, это описания, какие аппараты для чего нужны.

Мне нравится пример ксенобиологии. Внезапно, туда можно поместить таблицу слаймов с вики, потому что почему нет?
Ксенобиология же у нас филиал слаймренчера, а не отдел по исследованию неизвестных форм жизни. В особенности когда каждый второй ксенобиолог выводит слаймов по памяти, либо пользуется таблицей с вики.
Почему бы новичкам не упростить изучение ксенобиологии, добавив гайд прямо в игру, а тем, кто уже является адептом данной науки не добавить удобства в виде отсутствия необходимости лезть на вики?

Другой пример из этой же оперы - ксеноархеология. Опять же, личный опыт, но мне вот было крайне неудобно постоянно альттабаться, чтобы смотреть нужные вещи. Куда проще в книжке глянуть.

> А зачем им пилить гайды, когда уже есть "золотой стандарт" в виде официальных гайдов? Такие вещи очень демотивируют заниматься и без того не самым благодарным занятием. Что нужно сделать, так это завлечь игроков к тому, чтобы они делились опытом. Провести какой-нибудь конкурс или вроде того. Но никак не заменять всё бездушным копированием вики.

Чем эти гайды от коммьюнити лучше, чем "официальные" гайды с вики?
Кроме того, а почему бы не использовать удобную систему от ОП-а для интеграции гайдов от игроков? 

> Что нужно сделать, так это завлечь игроков к тому, чтобы они делились опытом. 

Как я уже писал в стене текста, игроки _уже_ делятся опытом друг с другом. Не знаю как остальные, но я вот каждый раунд, когда вижу в инженерном отделе как кто-то, кто делает что-то не по известным мне стандартам, спрашиваю, что он делает.
И не раз я брал небольшую команду из ассистентов (3 человека, считая меня) за инженера и показывал им разные вещи, связанные с инженерией (а также заставлял таскать баки с фороном). 
И во многих сферах так, новичков принимают довольно радушно в большинстве мест, по, опять же, моему опыту.
Люди хотят делиться опытом, это видно.
Но даже если я - человек, который хочет делиться опытом, то гайды я бы писать не стал. Почему?
Проблема, мне кажется, в том, что гайды, тем более ИЦ-гайды должны быть в крайне формальном стиле, сдерживаемом в общении. И, к тому же, намного сложнее донести суть без возможности сделать ООЦ-помарок в ЛООКе, например.
Передавать знания, отличные от базовых гайдами дико сложно из-за особенностей билда (🐛).

pushpushmergerebase commented:
В кои-то веки на ониксе завезли хорошую, красивую, грамотную фичу, а Мемори залупается. Не надо так.

Pos13 commented:
@MemoryDreams 
Я добавлю от себя несколько пояснений:
> Идеальным решением было бы склонить игроков к написанию собственных гайдов, сильно ограниченных по полноте информации, но дающие разные перспективы на разные проблемы. Унифицированные гайды убирают всякую нужду в разбирательстве что и как работает.

Теперь внедрить свой собственный гайд станет гораздо проще. Раньше для этого нужно было сделать форк, создать собственную книгу в коде и сделать Pull Request - вот это действительно _"демотивало заниматься и без того не самым благодарным занятием"_. Теперь можно будет просто создать/отредактировать статью на вики. Ты даже можешь заняться этим сам, все инструменты готовы и описаны в шапке. Если не нравится "унифицированный" гайд, можно исключить его из категории доступных для игры статей, а взамен включить другой, "урезанный".

> Внутреигровые книги и гайды на вики написаны на разных языках. Их нельзя мешать. Это будет ломать атмосферу.

Поздравляю! Теперь они будут написаны на одном языке и больше не будут ломать атмосферу!
Серьёзно, ты вообще заглядывал в эти старинные мануалы? Мало того, что эти кусочки английского текста были на сервере совершенно неуместны. Многие из них вдобавок были бессмысленны, неактуальны и практически бесполезны.

> Я считаю, что на вики должна быть только информация об управлении и всяких условностях, которые просто нельзя рассказать непосредственно в игре.

>  Но вот вопрос, если игрок вступает в раунд химиком или ксенобиологом, не зная, что гриндер нужен для измельчения твёрдых вещей, например, то возникает вопрос, а что он забыл на этой профессии?

Если новичку не хватает информации для игры, нужно ему помочь, а не бить ссанными тряпками.


И последнее - с интеграцией или без, игрок всегда может открыть вики в браузере во время игры. Мы не можем это ему запретить. Но мы можем создать условия, чтобы это не требовалось. Чтобы воспользоваться внитриигровой книгой было проще и интереснее. А уже над содержанием книг у нас будет полный контроль.

kokokobold commented:
Вещь хорошая, одобряю.

PlinHost commented:
> Давайте сделаем нормальные гайды, нв которые игроки смогут ссылаться в диалогах, а не вики странички, которые никто и никогда не будет открывать в игре.

Одно другому не мешает. Фича не заставляет использовать только гайды с вики, их можно вообще отключить.

PlinHost commented on code/modules/library/lib_items.dm:
```diff
@@ -53,10 +53,14 @@
 	if(contents.len)
-		var/obj/item/weapon/book/choice = input("Which book would you like to remove from the shelf?") as null|obj in contents
-		if(choice)
+		var/titles[length(contents)]
+		for(var/i in 1 to length(contents))
+			titles[i] = (istype(contents[i], /obj/item/weapon/book) && contents[i].title) ? contents[i].title : contents[i].name
+		var/title = input("Which book would you like to remove from the shelf?") as null|anything in titles
+		if(title)
 			if(!usr.canmove || usr.stat || usr.restrained() || !in_range(loc, usr))
 				return
+			var/obj/choice = contents[titles.Find(title)]
```
А нельзя здесь использовать ассоциативный список вместо индексов? Индекс может, например, устареть, пока ожидается `input`. И, я думаю, доставать должно быть можно только книги.

Pos13 replied:
~~Ассоциативный список нельзя, на полке может быть несколько книг с одинаковым названием.~~ Я не смог заставить input() выводить одинаковые названия, поэтому всё равно сделал ассоциативный список. Добавил проверку на случай, если полку сожрёт Нар-Си, пока игрок выбирает, что почитать.
> И, я думаю, доставать должно быть можно только книги.

Мало ли, что там потомки выдумают. Может, в contents будет хранится контрабанда.

PlinHost commented on code/modules/library/lib_items.dm:
```diff
@@ -224,12 +228,78 @@
+
+/obj/item/weapon/book/wiki/template
+	icon_state = "paper_stack"
+	desc = "Bunch of unbinded paper pieces."
```
```suggestion
	desc = "Bunch of unbound paper pieces."
```

PlinHost commented on code/modules/library/lib_machines.dm:
```diff
@@ -67,17 +67,44 @@
+				var/obj/item/weapon/book/wiki/template/template = O
+				print_wiki(template.topic, template.censored)
+			qdel(O)
+		else
+			..()
```
Почему родитель вызывается только если `operable()`? Он сам может проверить, если нужно.

Pos13 replied:
Исправил.

PlinHost commented on code/modules/modular_computers/file_system/programs/generic/library.dm:
```diff
@@ -227,3 +223,75 @@ The answer was five and a half years -ZeroBits
+		for(var/d in GLOB.cardinal)
+			var/obj/machinery/bookbinder/bndr = locate(/obj/machinery/bookbinder, get_step(nano_host(), d))
+			if(bndr && bndr.anchored && bndr.operable())
+				bndr.print_wiki(href_list["topic"], emagged ? 0 : 1)
```
В топик могут напихать любой фигни, и я даже не знаю, как его проверить, потому что на сервере нет списка валидных статей.

Pos13 replied:
Раскомментировал проверку на шаблон {{Ingame}} в JS. Теперь попытка открыть статью без шаблона будет заканчиваться вот таким сообщением: 
![image](https://user-images.githubusercontent.com/64479446/88035118-54dcdb00-cb4a-11ea-88fe-4704f6f1f0a6.png)
А попытка открыть несуществующую статью - сообщением "The page you specified doesn't exist".

PlinHost commented on config/example/config.txt:
```diff
@@ -564,7 +564,7 @@ DEFAULT_NO_VOTE
 
 ## Wiki URL
-# WIKIURL http://example.com
+WIKIURL https://wiki.ss13.ru
```
Лучше закомментированным оставить, мы же остальные юрл в примере конфига на свои не меняем.

Pos13 replied:
Исправил. Главное, чтобы на сервере не забыли поменять.

PlinHost reviewed (commented):
Жс, карту и спрайты не смотрел.

Pos13 committed:
Update code/modules/library/lib_items.dm

unbinded -> unbound

Co-authored-by: PlinHost <PlinHost@users.noreply.github.com>

Pos13 committed:
update(book_wiki): disallow books without {{Ingame}}

Статьи без шаблона {{Ingame}} теперь не будут показываться.
+ мелкие правки

Pos13 committed:
mapmerge(exodus-1, exodus-2)

Pos13 commented:
Я смержил карты, но гит всё ещё говорит, что нет.

Pos13 commented:
После мапмержа теперь не получается скомпилировать билд, т. к. на карте есть объекты, отсутствующие в этом ПР.

Pos13 committed:
merge(dev)

Pos13 committed:
Revert "merge(dev)"

This reverts commit ac308f008fd3badb6bb29eaa5cd0fc6343486468.

Pos13 committed:
Revert "mapmerge(exodus-1, exodus-2)"

This reverts commit aac14ee55eeeb0b6f3efa1c9819c1d4b364e4470.

Pos13 committed:
Merge branch 'dev' into library

Pos13 commented:
Конфликт карт разрешён.

Pos13 committed:
mapmerge #2

Pos13 committed:
Revert "mapmerge #2"

This reverts commit 1011f83d5973a27e4cc0f7815de9a4a88ad55ca8.

Pos13 committed:
Merge branch 'dev' into library

Pos13 committed:
mapmerge

One more time! Para-ra, pa-pa...

Pos13 committed:
step_x -> pixel_x
